File: engine/src/handwritten/interpolation.rs
use crate::interpolation::*;
use crate::presentation;
use crate::scope::Scope;
use borger_plugin_sdk::primitive::usize32;
use borger_plugin_sdk::traits::InterpolateTicks;

pub type Client = Scope<ClientOwned, ClientRemote>;

impl InterpolateTicks for presentation::Client {
	type InterpolationOutput = Client;
	fn interpolate_and_diff(
		prv: Option<&Self>,
		cur: &Self,
		amount: f32,
		received_new_tick: bool,
	) -> Self::InterpolationOutput {
		match cur {
			Self::Owned(cur) => Client::Owned(InterpolateTicks::interpolate_and_diff(
				prv.map(|prv| prv.as_owned().unwrap()),
				cur,
				amount,
				received_new_tick,
			)),
			Self::Remote(cur) => {
				if let Some(presentation::Client::Owned(prv)) = prv {
					//special case: received a new client id after reconnecting
					//to server. the previously owned client is now a stale
					//remote client that the server hasn't timed out yet
					Client::Remote(presentation::ClientRemote::interpolate_and_diff(
						Some(prv),
						cur,
						amount,
						received_new_tick,
					))
				} else {
					Client::Remote(presentation::ClientRemote::interpolate_and_diff(
						prv.map(|prv| prv.as_remote().unwrap()),
						cur,
						amount,
						received_new_tick,
					))
				}
			}
		}
	}
}

pub struct InterpolationContext {
	pub local_client_id: usize32,
	pub output: InterpolationOutput,
}

File: code_generator/src/files/interpolation.ts
import { ENGINE_GENERATED_DIR, stateWarningBlock, VALID_TYPES } from "@borger/code_generator/common.ts";
import { interpolablePrimitiveTypeSchema, type PrimitiveType } from "@borger/code_generator/state_schema.ts";
import {
	getPresentationStructName,
	presentationStructFilter,
} from "@borger/code_generator/files/presentation.ts";
import { writeFileSync } from "fs";
import type { FlattenedOutput } from "@borger/code_generator/flatten.ts";

export function generateInterpolation(flattened: FlattenedOutput) {
	writeFileSync(
		`${ENGINE_GENERATED_DIR}/interpolation.rs`,
		`${stateWarningBlock()}

use crate::simulation;
use crate::presentation;
use borger_plugin_sdk::traits::{PresentTick, Interpolate, InterpolateTicks};

${VALID_TYPES}

${flattened.output
	.map((group) =>
		group
			.filter(presentationStructFilter)
			.map(function generateInterpolationStruct(struct) {
				const interpolationStructName = getInterpolationStructName(struct.name);

				return `#[allow(non_camel_case_types, private_interfaces)]
pub struct ${interpolationStructName}
{
${struct.fields
	.filter((field) => field.presentation)
	.map(function generateInterpolationStructFields({ name, outerType, typeKind, innerType }) {
		//yes i know this is vile
		let interpolationType;
		if (typeKind === "collection")
			interpolationType = `<<${outerType}<simulation::${innerType}> as PresentTick>::PresentationOutput as InterpolateTicks>::InterpolationOutput`;
		else if (typeKind === "plugin")
			interpolationType = `<<${outerType} as PresentTick>::PresentationOutput as InterpolateTicks>::InterpolationOutput`;
		else interpolationType = outerType;

		return `	pub ${name}: ${interpolationType},`;
	})
	.join("\n")}
}

${generateInterpolateTicksImpl(false)}${
					struct.clientKind === "Remote"
						? `

${generateInterpolateTicksImpl(true)}`
						: ""
				}`;

				function generateInterpolateTicksImpl(downgradeScope: boolean) {
					const presentationStructName = getPresentationStructName(struct.name);
					const downgradedName = presentationStructName.replace(/Remote$/, "Owned"); //only valid if downgradeScope true
					return `impl InterpolateTicks${downgradeScope ? `<presentation::${downgradedName}>` : ""} for presentation::${presentationStructName}
{
	type InterpolationOutput = ${interpolationStructName};
	fn interpolate_and_diff(_prv: Option<&${downgradeScope ? `presentation::${downgradedName}` : "Self"}>, _cur: &Self, _amount: f32, _received_new_tick: bool) -> Self::InterpolationOutput
	{
		Self::InterpolationOutput
		{
${struct.fields
	.filter((field) => field.presentation)
	.map(function generateInterpolationImpl({ name, outerType, presentation, typeKind }) {
		let interpolationGetter;
		if (typeKind !== "primitive")
			interpolationGetter = `InterpolateTicks::interpolate_and_diff
			(
				_prv.map(|prv| &prv.${name}),
				&_cur.${name},
				_amount,
				_received_new_tick
			)`;
		else if (
			presentation === "clone" ||
			!(interpolablePrimitiveTypeSchema.options as PrimitiveType[]).includes(outerType as PrimitiveType)
		)
			interpolationGetter = `_cur.${name}`;
		else
			interpolationGetter = `if let Some(prv) = _prv
			{
				${outerType}::interpolate(prv.${name}, _cur.${name}, _amount)
			}
			else
			{
				_cur.${name}
			}`;

		return `			${name}: ${interpolationGetter},`;
	})
	.join("\n\n")}
		}
	}
}`;
				}
			})
			.join("\n\n"),
	)
	.join("\n\n")}
`,
	);
}

function getInterpolationStructName(simStructName: string) {
	return simStructName === "State" ? "InterpolationOutput" : simStructName;
}
